client/my-sites/plans-features-main/index.tsx

```diff
@@ -302,8 +308,20 @@ const PlansFeaturesMain = ( {
 			}
 		}
 
+		const selectedStorageOption =
+			selectedStorageOptions && selectedStorageOptions[ cartItemForPlan?.product_slug || '' ];
+		const storageAddOn = storageAddOns?.find( ( addOn ) => {
+			return addOn?.featureSlugs?.includes( selectedStorageOption || '' );
+		} );
+
+		const storageAddOnItemForPlan = storageAddOn && {
+			product_slug: storageAddOn.productSlug,
+			quantity: storageAddOn.quantity,
+			volume: 1,
+		};
+
 		if ( onUpgradeClick ) {
-			onUpgradeClick( cartItemForPlan );
+			onUpgradeClick( cartItemForPlan, storageAddOnItemForPlan );
```

Comment:
Good observation -- we'd have to update a relatively foundational method in order to handle multiple cart items at once ( [buildUpgradeFunction](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/signup/step-actions/index.js#L1273-L1284) ). This will improve code quality. From what I can tell, the surface are of code changes would also be small. It will, however, probably require extensively testing each signup flow to make sure everything continues to function as expected

We'd want to update:

`onUpgradeClick` -> `onSelectPlan` -> `buildUpgradeFunction`

@southp just wanted to keep you in the loop on this given our deadlines. If we move forward, it will probably delay delivery by a day or two ( to account for extra testing and code review ). The alternative is to create a follow-up, but I'm more than happy to work on this if we're all in agreement 🙂

Reply:
It's a valid concern worth addressing. From the other exsiting step action [addAddOnsToCart()](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/signup/step-actions/index.js#L638) we clearly need to make the data pipeline filling the shopping cart capable of handling multiple items. Let's give it a shot and I'll be happy to assist in testing and buying us more time if needed. 1~3 days doesn't sound too bad for me given what we'd gain from here :)

Reply:
Cool. Happy to assist with anything needed too /review, testing, etc. If we can unify the concept without unhealthy dose of complexity or side effects (like querying plans-ui store outside of plans code), then it sounds like good direction :-)

Reply:
Wonderful! Thanks for reaching a consensus with me everyone -- I believe I'll have something to address the cart items shortly 😃

Reply:
Hey folks! You might see checkout code being pushed in https://github.com/Automattic/wp-calypso/pull/81340/commits/027e4255bbb7fca37b4366ad3ec42947eeea5d03 -- the functionality appears to work, but to be clear, I'm still tidying things up 🙂